con-sdl.c

```diff
@@ -54,8 +54,8 @@ static void sdl_drawbitmap(int c, int x, int y)
     int bitcount = 0;
 	unsigned short bitvalue = 0;
 	int height = CHAR_HEIGHT;
-	char fg = (c & ATTR_FGCOLOR)? 255: 0;	// bright white fg for now
-	char bg = (c & ATTR_BGCOLOR)? 192: 0;	// medium white bg on reverse video
+	unsigned char fg = (c & ATTR_FGCOLOR)? 255: 0;	// bright white fg for now
+	unsigned char bg = (c & ATTR_BGCOLOR)? 192: 0;	// medium white bg on reverse video
```

Comment:
If the default video mode is 7 (monochrome 80x25), the character attribute should be decoded as the MDA does, except for the blinking. See https://www.seasip.info/VintagePC/mda.html.

Reply:
I'm not quite following you here: the character attribute is coded as the MDA (and EGA) does: 0x07. The values of 255 and 192 above are RGB values for the SDL display. They will be further extended when support for EGA color text attributes is added.

Reply:
This could be changed once the entire project is evaluated for a 2nd parameter to be used for attributes.
Again, its a bit complicated because the cursor code is dependent on OR'ing in a value to code point, so would like to defer to another PR, please.

Reply:
Here is a routine to compute the `fg` and `bg` intensities from the MDA character attributes:
```
// Convert character attribute according to video mode
// Currently only mode 7 (MDA) is implemented (without underlining or blinking)

#define MDA_BRIGHT 0xFF
#define MDA_NORMAL 0x7F
#define MDA_DARK   0x3F

static void attr_mda (byte_t attr, int * fg, int * bg)
	{
	switch (attr)
		{
		// Attributes 00h, 08h, 80h and 88h display as black space

		case 0x00:
		case 0x08:
		case 0x80:
		case 0x88:
			*fg = 0;
			*bg = 0;
			break;

		// Attribute 70h displays as black on green

		case 0x70: *fg = 0; *bg = MDA_NORMAL; break;

		// Attribute 78h displays as dark green on green

		case 0x78: *fg = MDA_DARK; *bg = MDA_NORMAL; break;

		// Attribute F0h displays as a blinking version of 70h (if blinking is enabled)
		// or as black on bright green otherwise

		case 0xF0: *fg = 0; *bg = MDA_BRIGHT; break;

		// Attribute F8h displays as a blinking version of 78h (if blinking is enabled)
		// as dark green on bright green otherwise

		case 0xF8: *fg = MDA_DARK; *bg = MDA_BRIGHT; break;

		// Bits 0-2: 1 = underline, other values = no underline
		// Bit 3: high intensity
		// Bit 7: blink

		default:
			*fg = (attr & 0x08) ? MDA_BRIGHT : MDA_NORMAL;
			*bg = 0;
		}
	}
```

Reply:
Hello @mfld-fr,

> Here is a routine to compute the fg and bg intensities from the MDA character attributes

Thanks! That could be added when we add MDA support. (Remember, EMU86 is currently only emulating EGA text mode at B800h at this point).  

The conversion from EGA/MDA attribute bytes to SDL 32-bit ARGB pixel format is best handled by a 16-entry RGB table lookup using three bits foreground or three bits background, along with the 'bright' bit. The reasons are many, but for instance the above proposed routine doesn't properly handle all 255 variations of the MDA attribute byte (which are produced by ELKS by color applications, see below). Also, you may want to actually display green on black when emulating MDA, which would be very easily done by substituting a seperate RGB color mapping table for the EGA one when MDA is emulating.

An 16-entry RGB mapping table also solves the minor, but important issue of the fact that using RGB 127,127,127 (0x7f) for normal 'white' is too dim, when the rest of the computer system is using 255,255,255 (0xff) for white (and also allows showing brighter mono/color). That's why 192 is used above for the ARGB value (which ends up being [192,192,192,255] in the SDL framebuffer).

If you could commit this, I'll add the EGA color support to show you what I mean, and then you can add MDA support (or rewrite it if you don't like it).

When you add MDA support, you'll have to set BDA 0040:0049 to 7 (we are running in mode 3 now), which tells the ELKS Direct Console to use segment B000h for the display buffer.  ELKS BIOS Console does not change any attribute values when using MDA vs EGA, so EGA "color" attribute mappings have to be carefully managed when displayed on real or emulated MDA. (The Compaq Portable does this when emulating EGA on its monochromatic display, another reason for an RGB lookup table). The routines in mem-io-elks.c using VID_BASE will have be updated as well. I would prefer to finish with the EGA emulation and few small cleanups before you/we add MDA, though.

Thank you!

Reply:
OK, thanks for the explanations, I did not catch that ELKS uses character attributes only in video mode 3 and not in 7. So I agree there is no need at that time for MDA attributes emulation.

I will change the default video mode from 7 to 3 in the PC/XT/AT BIOS stub for consistency, and put a 'unsupported attributes' warning when other OS would ask for mode 7.